File: samples/MiscellaneousSamples/app/src/com/pdftron/android/pdfnetsdksamples/util/LoggingOutputListener.java
package com.pdftron.android.pdfnetsdksamples.util;

import android.widget.ScrollView;
import android.widget.TextView;

import com.pdftron.android.pdfnetsdksamples.OutputListener;

import java.util.Arrays;

public class LoggingOutputListener implements OutputListener {

    private static final boolean SHOULD_THROW = false; // flag is used for testing purposes
    private final TextView mOutputTextView;
    private final ScrollView mOutputScrollView;

    public LoggingOutputListener(TextView outputTextView, ScrollView outputScrollView) {
        mOutputTextView = outputTextView;
        mOutputScrollView = outputScrollView;
    }

    @Override
    public void print(String output) {
        mOutputTextView.append(output);

        mOutputScrollView.post(new Runnable() {
            // This is necessary to scroll the ScrollView to the bottom.
            @Override
            public void run() {
                mOutputScrollView.fullScroll(ScrollView.FOCUS_DOWN);
            }
        });
    }

    @Override
    public void println(String output) {
        System.out.println(output);
        mOutputTextView.append(output + "\n");

        mOutputScrollView.post(new Runnable() {
            // This is necessary to scroll the ScrollView to the bottom.
            @Override
            public void run() {
                mOutputScrollView.fullScroll(ScrollView.FOCUS_DOWN);
            }
        });
    }

    @Override
    public void print() {
        print("");
    }

    @Override
    public void println() {
        println("");
    }

    @Override
    public void printError(String errorMessage) {
        if (SHOULD_THROW) {
            throw new RuntimeException(errorMessage);
        }
        println(errorMessage);
    }

    @Override
    public void printError(StackTraceElement[] stackTrace) {
        if (SHOULD_THROW) {
            throw new RuntimeException(Arrays.toString(stackTrace));
        }
        for (int i = 0; i < stackTrace.length; i++) {
            println(stackTrace[i].toString());
        }
    }
}

File: samples/MiscellaneousSamplesKotlin/app/src/com/pdftron/android/pdfnetsdksamples/util/LoggingOutputListener.kt
package com.pdftron.android.pdfnetsdksamples.util

import android.widget.ScrollView
import android.widget.TextView
import com.pdftron.android.pdfnetsdksamples.OutputListener
import java.util.*

class LoggingOutputListener(private val mOutputTextView: TextView, private val mOutputScrollView: ScrollView) : OutputListener {

    override fun print(output: String?) {
        mOutputTextView!!.append(output)

        mOutputScrollView!!.post {
            mOutputScrollView!!.fullScroll(ScrollView.FOCUS_DOWN)
        }

    }

    override fun println(output: String?) {
        System.out.println(output)
        mOutputTextView!!.append(output + "\n")

        mOutputScrollView!!.post {
            mOutputScrollView!!.fullScroll(ScrollView.FOCUS_DOWN)
        }

    }

    override fun print() {
        print("")
    }

    override fun println() {
        println("")
    }

    override fun printError(errorMessage: String) {
        if (SHOULD_THROW) {
            throw RuntimeException(errorMessage)
        }
        println(errorMessage)
    }

    override fun printError(stackTrace: Array<StackTraceElement>) {
        if (SHOULD_THROW) {
            throw RuntimeException(Arrays.toString(stackTrace))
        }
        for (i in stackTrace.indices) {
            println(stackTrace[i].toString())
        }
    }

    companion object {

        private val SHOULD_THROW = false // flag is used for testing purposes
    }
}
